pkg/blockbuilder/builder/slimgester.go

```diff
@@ -95,94 +111,170 @@ type BlockBuilder struct {
 	}
 
 	i.Service = services.NewBasicService(nil, i.running, nil)
 	return i, nil
 }
 
 func (i *BlockBuilder) running(ctx context.Context) error {
-	ticker := time.NewTicker(i.cfg.Interval)
-	defer ticker.Stop()
+	wg := sync.WaitGroup{}
 
-	// run once in beginning
-	select {
-	case <-ctx.Done():
-		return nil
-	default:
-		_, err := i.runOne(ctx)
-		if err != nil {
-			level.Error(i.logger).Log("msg", "block builder run failed", "err", err)
-		}
+	for j := 0; j < i.cfg.WorkerParallelism; j++ {
+		wg.Add(1)
+		go func(id string) {
+			defer wg.Done()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					err := i.runOne(ctx, id)
+					if err != nil {
+						level.Error(i.logger).Log("msg", "block builder run failed", "err", err)
+					}
+				}
+			}
+		}(fmt.Sprintf("worker-%d", j))
 	}
 
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case <-ticker.C:
-			skipped, err := i.runOne(ctx)
-			level.Info(i.logger).Log(
-				"msg", "completed block builder run", "skipped",
-				"skipped", skipped,
-				"err", err,
-			)
-			if err != nil {
-				level.Error(i.logger).Log("msg", "block builder run failed", "err", err)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		ticker := time.NewTicker(i.cfg.SyncInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				if err := i.syncJobs(ctx); err != nil {
+					level.Error(i.logger).Log("msg", "failed to sync jobs", "err", err)
+				}
 			}
 		}
-	}
+	}()
+
+	wg.Wait()
+	return nil
 }
 
-// runOne performs a single
-func (i *BlockBuilder) runOne(ctx context.Context) (skipped bool, err error) {
+func (i *BlockBuilder) syncJobs(ctx context.Context) error {
+	i.jobsMtx.RLock()
+	defer i.jobsMtx.RUnlock()
+
+	for _, job := range i.inflightJobs {
+		if err := i.transport.SendSyncJob(ctx, &types.SyncJobRequest{
+			BuilderID: i.id,
+			Job:       job,
+		}); err != nil {
+			level.Error(i.logger).Log("msg", "failed to sync job", "err", err)
+		}
+	}
+
+	return nil
+}
 
-	exists, job, err := i.jobController.LoadJob(ctx)
+func (i *BlockBuilder) runOne(ctx context.Context, workerID string) error {
+	// assuming GetJob blocks/polls until a job is available
+	resp, err := i.transport.SendGetJobRequest(ctx, &types.GetJobRequest{
+		BuilderID: workerID,
+	})
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	if !exists {
+	if !resp.OK {
 		level.Info(i.logger).Log("msg", "no available job to process")
-		return true, nil
+		return nil
 	}
 
+	job := resp.Job
 	logger := log.With(
 		i.logger,
+		"worker_id", workerID,
 		"partition", job.Partition,
 		"job_min_offset", job.Offsets.Min,
 		"job_max_offset", job.Offsets.Max,
 	)
 
+	// TODO: add metrics for inflight jobs
+	i.jobsMtx.Lock()
+	i.inflightJobs[job.ID] = job
+	i.jobsMtx.Unlock()
+
+	lastConsumedOffset, err := i.processJob(ctx, job, logger)
+	// TODO: pass lastConsumedOffset as a separate field
+	job.Offsets.Max = lastConsumedOffset
```

Comment:
I think it's much simpler if the jobs are predetermined at the scheduler. This was the initial design and although it does introduce a bit of lag (we only process offsets known at the time of job creation), I think the simplicity & separation of concerns are more beneficial (at least for now).